feat: specify how to install GZIP-compressed modules

roman-kashitsyn opened the pull request:
The system now supports multiple encodings of the `wasm_module` field: raw Wasm and gzip-compressed Wasm.
We pick the appropriate decoding based on the magic number of the binary.

nomeata commented:
Did you consider using the magic byte of the uploaded file to determine the format? They are there for that purpose, and easy and reliable to use, so I am not 100% sure it’s worth it to make the Candid interface more complex with such redundant information. But also don’t feel strongly about it.

Dfinity-Bjoern commented:
> So if the goal is to add compression, maybe we should build that in a way that later extends to such a future? I.e. don’t use `gzip` but rather `zip` and say that that it has to contain a single file with a file name ending in `.wasm`? This would mostly achieve what the gzip-compression achieve, but be forward-compatible to richer schemes.

I agree in principle, but I also think that we haven't understood that future at this point, and the simple `gzip` compression addresses an important pain point right now. So I do think it makes sense to go ahead with `gzip` compression for now. Before making full use of an archive upload, we probably want to allow for larger ingress messages.

I do agree that the format-related description should be moved to the "module format" section, and I'd be eager to hear @roman-kashitsyn 's opinion on the magic byte because also to me it seems the natural solution.

roman-kashitsyn commented on spec/index.adoc:
```diff
@@ -1587,6 +1587,14 @@ This is atomic: If the response to this request is a `reject`, then this call ha
+
+* If the `wasm_module_encoding` is `null` or `opt identity`, the system parses `wasm_module` as raw WebAssembly binary.
+* If the `wasm_module_encoding` is `opt gzip`, the system decodes the contents of `wasm_module` using the https://en.wikipedia.org/wiki/LZ77_and_LZ78#LZ77[Lempel-Ziv algorithm (LZ77)] before parsing the module.
+====
```
Done

roman-kashitsyn commented:
> Did you consider using the magic byte of the uploaded file to determine the format?

Yep, @crusso suggested this during the feature review and I liked this idea a lot. I implemented this in a prototype, this simplifies a lot of things. I changed the spec accordingly.

> So if the goal is to add compression, maybe we should build that in a way that later extends to such a future? I.e. don’t use gzip but rather zip and say that that it has to contain a single file with a file name ending in .wasm? This would mostly achieve what the gzip-compression achieve, but be forward-compatible to richer schemes.

In theory, gzip also supports multiple files in a single stream (though tarballs compress better), so we could extend the feature later without breaking backward compatibility.
Archive uploads are fine, but I think it's too early to specify such a complex feature.

Dfinity-Bjoern reviewed (approved):
Thanks @roman-kashitsyn!
